feat: add jwt pkg

sreya opened the pull request:
This PR adds a `jwtutils` package to be shared amongst the various packages in the codebase that make use of JWTs. It's intended to help us standardize on one library instead of some implementations using `go-jose` and others using `golang-jwt`.

The main reason we're converging on `go-jose` is due to its support for JWEs, `golang-jwt` also has a repo to handle it but it [doesn't look maintained](https://github.com/golang-jwt/jwe)

sreya committed:
feat: add jwt pkg

sreya committed:
update make gen

sreya committed:
Refactor JWT package to modularize key functions

This refactor encapsulates key management for JWT signing and verification by:
- Introducing `ParseKeyFunc` and `SecuringKeyFn` to streamline key operations.
- Moving key-related functions to `jwt.go` to improve package modularity.
- Simplifying `Encrypt` and `Decrypt` function documentation and usage.
- Updating tests to align with the new function signatures and logic flow.

sreya committed:
Remove unused JWT test file from repository

The JWT functionality has been refactored, and its associated test values are no longer required. Removing the test file helps maintain code clarity and prevents outdated test logic from impacting the current codebase.

sreya committed:
Refactor JWT key functions and add tests

- Renamed key functions to clarify their use.
- Removed unused `featureHeaderKey`.
- Added comprehensive tests for JWT signing and verifying.

sreya committed:
Rename VerifyFn to ParseFn in JWT tests

sreya committed:
Remove unused JWE test file

sreya committed:
Refactor JWT test structs to use public field names

deansheather commented on coderd/jwt/jwe.go:
```diff
@@ -0,0 +1,102 @@
+package jwt
```
Maybe this package should be renamed to avoid conflicting with jose's jwt package name. We will need to change the import name for jose/jwt everywhere we use this

sreya commented on coderd/jwt/jwt.go:
```diff
@@ -0,0 +1,79 @@
+type SecuringKeyFn func() (id string, key interface{}, err error)
+
+// SecuringKeyFromCache returns the appropriate key for signing or encrypting.
+func SecuringKeyFromCache(ctx context.Context, keys cryptokeys.Keycache) SecuringKeyFn {
```
It needs to work with `coderd/license` and I'd rather not force a caller to use the `Keycache` abstraction if it's not related to key rotation.

deansheather replied:
If you use an interface couldn't the license package just implement it with a static key?

sreya replied:
Yes but I'm trying to avoid leaking `cryptokeys` into packages that have nothing to do with key rotation

deansheather replied:
TBH there's at least 3 places that will use cryptokeys and only one that won't so I think the API should optimize for the more common case. The interfaces `SecuringKeyProvider` and `VerifyingKeyProvider` could just go in this package instead and only needs to implement the single methods necessary for each use

sreya replied:
I originally had it simply accept a `Keycache` but I removed in favor of an API that more closely resembles the `golang-jwt` API and had a more general-purpose feel to it but I can return to that if we think that's better. 

My only concern is if something doesn't fit neatly into the interface in the future then you have no choice but to either completely refactor the package or sidestep it and duplicate a lot of code.

sreya replied:
Updated based on this feedback fwiw

sreya commented on coderd/jwt/jwe.go:
```diff
@@ -0,0 +1,102 @@
+)
+
+// Encrypt encrypts a token and returns it as a string.
+func Encrypt(claims Claims, keyFn SecuringKeyFn) (string, error) {
```
> should check for issuing keys with missing expiry dates

I'm not sure how I could do that without reflection 🤔

sreya committed:
Refactor JWT to use new crypto key management system

- Streamlines key retrieval by integrating with the cryptokeys package.
- Simplifies key handling by decoding directly using hex utilities.
- Enhances validation by using context for managing key operations.
- Improves test robustness with clearer setup and consistent mocking.

sreya committed:
Refactor JWT package for improved modularity and clarity

- Rename package `jwt` to `jwtutils` for better context.
- Consolidate constants and enhance error messages.
- Simplify key management for JWT signing and verification.

sreya committed:
mv dir

sreya committed:
update references

deansheather commented on coderd/jwtutils/jwe.go:
```diff
@@ -0,0 +1,127 @@
+)
+
+// Encrypt encrypts a token and returns it as a string.
+func Encrypt(ctx context.Context, keys cryptokeys.Keycache, claims Claims) (string, error) {
```
Could this accept an interface instead of the cryptokeys type? It only needs the Signing method

deansheather commented on coderd/jwtutils/jwe.go:
```diff
@@ -0,0 +1,127 @@
+
+// Encrypt encrypts a token and returns it as a string.
+func Encrypt(ctx context.Context, keys cryptokeys.Keycache, claims Claims) (string, error) {
+	signing, err := keys.Signing(ctx)
```
One issue with this system is that you could call Encrypt with a key that's only meant for signing. It'd be nice if this method was `Encrypting` instead, and the underlying keycache rejected with an error if the key is being used for the wrong purpose

deansheather commented on coderd/jwtutils/jwe.go:
```diff
@@ -0,0 +1,127 @@
+		return xerrors.Errorf("parse sequence: %w", err)
+	}
+
+	key, err := keys.Verifying(ctx, int32(sequence))
```
Same here but `Decrypting`

deansheather commented on coderd/jwtutils/jwt.go:
```diff
@@ -0,0 +1,28 @@
+}
+
+// ParseOptions are options for parsing a JWT.
+type ParseOptions struct {
```
TBH these could be two separate types to make it harder to use incorrectly. I don't think any real code benefits from these being the same except for the test code.

sreya committed:
refactor interfaces

sreya committed:
refactor dbkeycache

sreya committed:
Refactor JWT utility options for flexibility

Refactor JWT utility code to replace `ParseOptions` with more specific
`VerifyOptions` and `DecryptOptions`. This change enhances clarity and
flexibility, accommodating different needs for JWS and JWE operations. Update
tests to align with the refactored structure, ensuring robust coverage and
functionality verification.

sreya committed:
Enhance key generation and JWT error messages

Increase OIDC secret length for added security and clarity.
Optimize JWT test coverage with custom claims and key cache tests.

sreya committed:
Update cryptographic key length requirements

sreya committed:
Refactor key provider interfaces in JWT utilities

sreya committed:
Refactor dbCache to remove feature validation

Emyrk commented on coderd/cryptokeys/dbkeycache.go:
```diff
@@ -34,18 +33,34 @@ type DBCache struct {
 // release resources associated with its internal timer.
-func NewDBCache(logger slog.Logger, db database.Store, feature database.CryptoKeyFeature, opts ...func(*DBCache)) *DBCache {
-	d := &DBCache{
+func NewSigningCache(logger slog.Logger, db database.Store, feature database.CryptoKeyFeature, opts ...func(*dbCache)) (SigningKeycache, error) {
+	if !isSigningKeyFeature(feature) {
+		return nil, ErrInvalidFeature
+	}
+
+	return newDBCache(logger, db, feature, opts...), nil
+}
+
+func NewEncryptionCache(logger slog.Logger, db database.Store, feature database.CryptoKeyFeature, opts ...func(*dbCache)) (EncryptionKeycache, error) {
+	if !isEncryptionKeyFeature(feature) {
+		return nil, ErrInvalidFeature
+	}
+
+	return newDBCache(logger, db, feature, opts...), nil
+}
```
Returning `EncryptionKeycache` or `SigningKeycache` is a nice way to prevent the caller from encrypting & signing with the same key (which is a big no-no).

You could still typecast this into the other type. Do you think adding `isEncryptionKeyFeature` and `isSigningKeyFeature` to the functions that return the signing/encrypt key would be excessive? Just to futher protect from any misuse.

sreya replied:
I was doing the checks originally but I removed it because it feels like it's so outside the spirit of the API it's not even worth accounting for. I can add it back though if you think it's better.

Emyrk replied:
Your call, I felt the check was cheap

sreya replied:
Done

Emyrk commented on coderd/cryptokeys/dbkeycache.go:
```diff
@@ -34,18 +33,34 @@ type DBCache struct {
+	return newDBCache(logger, db, feature, opts...), nil
+}
+
+func newDBCache(logger slog.Logger, db database.Store, feature database.CryptoKeyFeature, opts ...func(*dbCache)) *dbCache {
+	d := &dbCache{
```
I do not think `d.timer = d.clock.AfterFunc(never, d.clear)` works as expected. The `d.clear` is never called.

https://go.dev/play/p/NrIUBz3Ea3J

sreya replied:
It's just a way to initialize the timer, it gets set to a proper value the first time a fetch to the DB is performed.

Emyrk replied:
Maybe comment above? Just felt weird to read

sreya replied:
Agreed it does look strange

Emyrk commented on coderd/cryptokeys/dbkeycache.go:
```diff
@@ -61,18 +76,39 @@ func NewDBCache(logger slog.Logger, db database.Store, feature database.CryptoKe
 }
 
-// Verifying returns the CryptoKey with the given sequence number, provided that
+func (d *dbCache) EncryptingKey(ctx context.Context) (id string, key interface{}, err error) {
+	return d.latest(ctx)
+}
+
+func (d *dbCache) DecryptingKey(ctx context.Context, id string) (key interface{}, err error) {
+	return d.sequence(ctx, id)
+}
+
+func (d *dbCache) SigningKey(ctx context.Context) (id string, key interface{}, err error) {
+	return d.latest(ctx)
+}
+
+func (d *dbCache) VerifyingKey(ctx context.Context, id string) (key interface{}, err error) {
+	return d.sequence(ctx, id)
+}
```
I know `jwtutils` is the user of this package, and handles the `interface{}`, but it is unfortunate to need to use `interface{}` :cry:.

I took a stab at throwing in generics to get type safety, but it locks in a feature to a specific algo.

sreya replied:
Yeah it is :(

Emyrk commented on coderd/cryptokeys/dbkeycache.go:
```diff
@@ -61,18 +76,39 @@ func NewDBCache(logger slog.Logger, db database.Store, feature database.CryptoKe
 }
 
-// Verifying returns the CryptoKey with the given sequence number, provided that
+func (d *dbCache) EncryptingKey(ctx context.Context) (id string, key interface{}, err error) {
+	return d.latest(ctx)
+}
+
+func (d *dbCache) DecryptingKey(ctx context.Context, id string) (key interface{}, err error) {
+	return d.sequence(ctx, id)
+}
```
`Encrypting` only returns the latest, and `Decrypting` returns the key by ID **iff** it's still valid?

Can we add comments to these functions, and to the Signing ones below to mention this behavior?

sreya replied:
Added to the interfaces

Emyrk commented on coderd/jwtutils/jwe.go:
```diff
@@ -0,0 +1,125 @@
+		return "", xerrors.Errorf("encrypt: %w", err)
+	}
+
+	serialized := []byte(encrypted.FullSerialize())
+	return base64.RawURLEncoding.EncodeToString(serialized), nil
```
Is there a reason you use `CompactSerialize` for JWS and `FullSerialized` + manual base64 encoding here?

The base64 encoding is not the compact base64 encoding, you can see the `ParseEncryptedCompact` here: https://github.com/go-jose/go-jose/blob/main/jwe.go#L286-L330

So there is no dots I do not think if we do this ourselves.

sreya replied:
Honestly I just copied the implementation verbatim but I agree I don't see why we shouldn't make the change.

Emyrk commented on coderd/jwtutils/jwe.go:
```diff
@@ -0,0 +1,125 @@
+		opt(&options)
+	}
+
+	encrypted, err := base64.RawURLEncoding.DecodeString(token)
+	if err != nil {
+		return xerrors.Errorf("decode: %w", err)
+	}
+
+	object, err := jose.ParseEncrypted(string(encrypted),
```
The inverse of above. If you do `base64.RawURLEncoding`, then you have to manually base64 it back into `json`. But we probably want the compact base64 format, then we do not need to use the `base64` package at all.

Emyrk commented on coderd/jwtutils/jws.go:
```diff
@@ -0,0 +1,129 @@
+type VerifyOptions struct {
+	RegisteredClaims jwt.Expected
+
+	// The following are only used for JWSs.
```
Can you use `VerifyOptions` for anything other than a JWS? The comment reads like these options can be used for something else.

Emyrk commented on coderd/jwtutils/jws.go:
```diff
@@ -0,0 +1,129 @@
+		return xerrors.Errorf("verify payload: %w", err)
+	}
+
+	err = json.Unmarshal(payload, &claims)
+	if err != nil {
+		return xerrors.Errorf("unmarshal payload: %w", err)
+	}
```
Is there a case we would not use `jwt.Claims`? Could we just have `Verify` and `Decrypt` return a `jwt.Claims`?

If there is another interface that fulfills the interface that we are using, then we could also do a generic method.

```golang
// Verify verifies that a token was signed by the provided key. It unmarshals into the provided claims.
func Verify[T Claims](ctx context.Context, v VerifyKeyProvider, token string, opts ...func(*VerifyOptions)) (Claims, error) {
	var empty T
	options := VerifyOptions{
		RegisteredClaims: jwt.Expected{
			Time: time.Now(),
		},
		SignatureAlgorithm: signingAlgo,
	}

	for _, opt := range opts {
		opt(&options)
	}

	object, err := jose.ParseSigned(token, []jose.SignatureAlgorithm{options.SignatureAlgorithm})
	if err != nil {
		return empty, xerrors.Errorf("parse JWS: %w", err)
	}
	
	// ...
	
	var claims T
	err = json.Unmarshal(payload, &claims)
	if err != nil {
		return empty, xerrors.Errorf("unmarshal payload: %w", err)
	}

	return claims, claims.Validate(options.RegisteredClaims)
}
```

But then you'd have to call like:
`claims, err = jwtutils.Verify[jwt.Claims](ctx, key, token, withVerifyExpected(c.expectedClaims))` which is kinda annoying :thinking: 


----

I'll let you decide.

sreya replied:
I believe we discussed this offline but you basically need to allow for an arbitrary payload. The `jwt.Claims` are default claims e.g. (`iss`, `sub`, etc)

sreya committed:
Refactor cryptokeys and jwtutils interfaces and logic

- Enhance comments for key interfaces to clarify usage
  and considerations for time validity and clock skew.
- Refactor JWE/JWS logic to simplify serialization and
  deserialization processes, ensuring more efficient and
  concise handling of JWTs. Implement compact serialization
  and remove unnecessary base64 encoding.

sreya committed:
Remove unused test code and mock cleanup

Removing unused test and mock code simplifies maintenance and reduces clutter.

sreya committed:
Remove cryptokeys keycachemock from Makefile

Deleting unnecessary cryptokeys keycachemock ensures clarity and reduces maintenance.

sreya committed:
Add feature validation to dbCache key methods

Ensure that the dbCache methods for encrypting, decrypting, signing,
and verifying keys validate the feature flag before proceeding with
operations. This validation step prevents using keys for unintended
purposes, maintaining proper alignment with their intended
cryptographic feature.

sreya committed:
fmt

Emyrk reviewed (approved):
LG, would like to have the "fetcher" interface so we can use this in wsproxy too at some point :+1:

sreya commented:
> LG, would like to have the "fetcher" interface so we can use this in wsproxy too at some point 👍

Yeah I'll open an issue to work on a refactor once I get this running in dev.

sreya committed:
Add initialization comment for db key cache timer
